cmd/posix-list-dir_unix.go

```diff
@@ -75,6 +75,60 @@ func readDir(dirPath string) (entries []string, err error) {
 	return readDirN(dirPath, -1)
 }
 
+// readDir applies the filter function on each entries at dirPath, doesn't recurse into
+// the directory itself.
+func readDirFilterFn(dirPath string, filter func(name string, typ os.FileMode) error) error {
+	fd, err := syscall.Open(dirPath, 0, 0)
+	if err != nil {
+		if os.IsNotExist(err) || isSysErrNotDir(err) {
+			return errFileNotFound
+		}
+		if os.IsPermission(err) {
+			return errFileAccessDenied
+		}
+		return err
+	}
+	defer syscall.Close(fd)
+
+	buf := make([]byte, blockSize) // stack-allocated; doesn't escape
+	boff := 0                      // starting read position in buf
+	nbuf := 0                      // end valid data in buf
+
+	for {
+		if boff >= nbuf {
+			boff = 0
+			nbuf, err = syscall.ReadDirent(fd, buf)
+			if err != nil {
+				if isSysErrNotDir(err) {
+					return errFileNotFound
+				}
+				return err
+			}
+			if nbuf <= 0 {
+				break
+			}
+		}
+		consumed, name, typ, err := parseDirEnt(buf[boff:nbuf])
```

Comment:
With this logic it assumes that `buf[boff:nbuf]` contains a full directory entry.

If it doesn't there is no protection against an infinite loop as far as I can tell.

Maybe a check `if consumed == 0 &&  boff < nbuf  { return internal error... }`

Reply:
> With this logic it assumes that `buf[boff:nbuf]` contains a full directory entry.

@klauspost yes 8KiB sufficient space, in-fact 4096 is enough 8192 is to handle a corner case on macOS - for any such entry that is the assumption here. This is the current code anyway so we have never seen any issue.

Reply:
It is a harmless check. No need to wait for it to potentially blow up IMO.

Reply:
Correct but parseDirEnt makes sure that consumed when is set to 0 there is an error always, otherwise it points to dirent.Reclen - so it is not possible to reach an infinite loop. 

We may be simply adding defensive code IMO, I can add but it doesn't add much value.

Reply:
For example inside parseDirEnt() makes sure such a thing never happens.

```
         if len(buf) < int(dirent.Reclen) {                                                                                                              
                 return consumed, name, typ, fmt.Errorf("buf size %d < record length %d", len(buf), dirent.Reclen)                                       
         } 
```